src/alignment/distance.rs

```diff
@@ -57,32 +58,39 @@ pub fn hamming(alpha: TextSlice<'_>, beta: TextSlice<'_>) -> u64 {
 /// let ldist = levenshtein(x, y); // Distance is 5
 /// assert_eq!(ldist, 5);
 /// ```
-#[allow(unused_assignments)]
+#[allow(non_snake_case)]
 pub fn levenshtein(alpha: TextSlice<'_>, beta: TextSlice<'_>) -> u32 {
-    let mut columns = [vec![0u32; alpha.len() + 1], vec![0u32; alpha.len() + 1]];
-    let mut i_prev = 0;
-    let mut i_cur = 1;
+    let (m, n) = (alpha.len(), beta.len());
+    let mut M: Vec<u32> = Vec::with_capacity(m + 1); // the dynamic programming matrix (only 1 column stored)
+    let mut s: u32; // M[i - 1][j - 1]
+    let mut c: u32; // M[i - 1][j]
+    let mut a: u8; // alpha[i]
+    let mut b: u8; // beta[j]
 
-    for i in 0..columns[0].len() {
-        columns[0][i] = i as u32;
+    // 0th column
+    for i in 0..=(m as u32) {
+        M.push(i);
     }
-
-    for (j, item) in beta.iter().enumerate() {
-        i_cur %= 2;
-        i_prev = 1 - i_cur;
-
-        columns[i_cur][0] = 1 + j as u32;
-        for i in 1..columns[0].len() {
-            columns[i_cur][i] = min(
-                columns[i_prev][i - 1] + if alpha[i - 1] == *item { 0 } else { 1 },
-                min(columns[i_cur][i - 1] + 1, columns[i_prev][i] + 1),
-            );
+    // columns 1 to n - 1
+    for j in 1..=n {
+        s = (j - 1) as u32;
+        c = j as u32;
+        b = unsafe { *beta.get_unchecked(j - 1) };
```

Comment:
I'd be interested if this actually makes a difference. The compiler already knows:

- length of `beta`
- possible values of `j`

so I think the resulting bytecode might be identical. Perhaps the `j - 1` makes it more difficult, but then again j could go from `0..=n-1` instead.

Reply:
Yes, big difference. Indeed, the speed improvement is mainly due to using `unsafe`, not halving the number of vectors.

Running `cargo bench levenshtein_dist` on https://github.com/rust-bio/rust-bio/pull/365/commits/adacb35001e4ab78a6e9a9723c1ac91458c89abb

```
test bench_levenshtein_dist_diverse_str              ... bench:  48,969,794 ns/iter (+/- 3,287,370)
test bench_levenshtein_dist_diverse_str_safe         ... bench:  69,967,289 ns/iter (+/- 10,004,630)
test bench_levenshtein_dist_equal_str                ... bench:  49,172,499 ns/iter (+/- 8,427,747)
test bench_levenshtein_dist_equal_str_safe           ... bench:  68,952,702 ns/iter (+/- 7,086,610)
test bench_levenshtein_dist_worst_str                ... bench:  48,885,604 ns/iter (+/- 4,377,159)
test bench_levenshtein_dist_worst_str_safe           ... bench:  72,058,646 ns/iter (+/- 9,063,159)
```

Reply:
Interesting, thank you for benchmarking!

Reply:
Probably it was `j-1` that makes it difficult. When I applied unsafe indexing to hamming distance, there was no significant improvement in speed.